### PR TITLE
chore(groupcache): Support per cache group capacity

### PR DESCRIPTION
**Pull request by kavirajk**
Signed-off-by: Kaviraj <kavirajkanagaraj@gmail.com>

**What this PR does / why we need it**:
Support per cache group capacity. 

With this individual group in groupcache can have own memory threshold.

**Which issue(s) this PR fixes**:
Fixes # https://github.com/grafana/loki/issues/6786

**Special notes for your reviewer**:
1. Will add per group ballast config in separate PR after testing
2. Will replace `/groupcache/` with generic name also in different PR.

**Checklist**
- [ ] Documentation added
- [x] Tests updated
- [ ] Is this an important fix or new feature? Add an entry in the `CHANGELOG.md`.
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/upgrading/_index.md`

### CODE REVIEW

File: pkg/loki/modules.go [hunk 1]
```diff
@@ -166,10 +166,16 @@ func (t *Loki) initGroupcache() (_ services.Service, err error) {
...
-
-	// The index cache generates too much traffic to be used. Make it a fifo cache
+	t.Cfg.ChunkStoreConfig.ChunkCacheConfig.GroupCache = gc.NewGroup(
+		t.Cfg.ChunkStoreConfig.ChunkCacheConfig.Prefix+"groupcache",
+		&t.Cfg.ChunkStoreConfig.ChunkCacheConfig.Groupcache,
+		stats.ChunkCache,
+	)
+	t.Cfg.QueryRange.ResultsCacheConfig.CacheConfig.GroupCache = gc.NewGroup(
+		t.Cfg.QueryRange.ResultsCacheConfig.CacheConfig.Prefix+"groupcache",
+		&t.Cfg.ChunkStoreConfig.ChunkCacheConfig.Groupcache,
```
Comment (dannykopping) [inline pkg/loki/modules.go:176]: Why the chunk cache config for the results cache?

File: pkg/storage/chunk/cache/cache.go [hunk 2]
```diff
@@ -35,6 +35,7 @@ type Config struct {
 	MemcacheClient MemcachedClientConfig `yaml:"memcached_client"`
 	Redis          RedisConfig           `yaml:"redis"`
 	Fifocache      FifoCacheConfig       `yaml:"fifocache"`
+	Groupcache     GroupConfig           `yaml:"groupcache"` // TODO(kavi): consider resolving conflict with `GroupCache` field in same struct.
```
Comment (dannykopping) [inline pkg/storage/chunk/cache/cache.go:38]: Is this TODO still relevant?

File: pkg/storage/chunk/cache/groupcache.go [hunk 3]
```diff
@@ -69,6 +69,11 @@ type GroupCacheConfig struct {
 	Cache Cache `yaml:"-"`
 }
 
+// GroupGconfig represents config per Group.
```
Comment (dannykopping) [inline pkg/storage/chunk/cache/groupcache.go:72]:
```suggestion
// GroupConfig represents config per Group.
```

File: pkg/storage/chunk/cache/groupcache.go [hunk 4]
```diff
@@ -198,15 +203,19 @@ func (c *GroupCache) Stats() *groupcache.Stats {
...
-	logger        log.Logger
-	wg            *sync.WaitGroup
-	cacheType     stats.CacheType
+	cache     *groupcache.Group
+	logger    log.Logger
+	wg        *sync.WaitGroup
+	cacheType stats.CacheType
+
+	// cacheBytes represents maxSize (in bytes) this group can grow.
+	cacheBytes int64 // TODO(kavi): expose it as _info metrics later?
```
Comment (dannykopping) [inline pkg/storage/chunk/cache/groupcache.go:212]: I think this metric would belong with the other groupcache metrics, but yes i love the idea of making this a metric

File: pkg/loki/modules.go [hunk 5]
```diff
@@ -166,10 +166,16 @@ func (t *Loki) initGroupcache() (_ services.Service, err error) {
 		return nil, err
 	}
 
-	t.Cfg.ChunkStoreConfig.ChunkCacheConfig.GroupCache = gc.NewGroup(t.Cfg.ChunkStoreConfig.ChunkCacheConfig.Prefix+"groupcache", stats.ChunkCache)
-	t.Cfg.QueryRange.ResultsCacheConfig.CacheConfig.GroupCache = gc.NewGroup(t.Cfg.QueryRange.ResultsCacheConfig.CacheConfig.Prefix+"groupcache", stats.ResultCache)
-
-	// The index cache generates too much traffic to be used. Make it a fifo cache
```
Comment (dannykopping) [inline pkg/loki/modules.go:172]: Let's not lose this comment

File: pkg/storage/chunk/cache/groupcache.go [hunk 6]
```diff
@@ -215,6 +224,11 @@ func (c *GroupCache) NewGroup(name string, ct stats.CacheType) Cache {
 	c.wg.Add(1)
 	c.startWaitingForClose()
 
+	cap := c.capacity
+	if cfg.CapacityMB != 0 {
+		cap = cfg.CapacityMB * 1e6 // MB into bytes
+	}
```
Comment (dannykopping) [inline pkg/storage/chunk/cache/groupcache.go:230]: Hhmm, do we need to keep the `capacity` field here?

Comment (kavirajk) [reply pkg/storage/chunk/cache/groupcache.go:230]: I would think so. `c.capacity` comes from global groupcache config (when we create single groupcache instance from `common.Groupcache` config).

`Groupcache.capacity` -> global capacity
`group.cacheBytes` -> local to particular group.

We need both.

does that answers your question? or is it something else?. 

nit: I can make both consistent name as `cacheBytes` (in sync with how it's used in original google/mailgun Groupcache package).

File: pkg/loki/modules.go (hunk 1 above)
Comment (kavirajk) [reply pkg/loki/modules.go]: good catch. Thanks :+1: . Fixed. (was bit haste while resolving merge conflict with Travis PR)

File: pkg/storage/chunk/cache/cache.go (hunk 2 above)
Comment (kavirajk) [reply pkg/storage/chunk/cache/cache.go:38]: Yes. we have two fields on this struct.

`Groupcache` -> GroupConfig (named it this way to be consistent with fifocache)
`GroupCache` -> Instance of groupcache we populate in `init modules.go`. (not exposed `yaml:"-"`)

Any suggestions welcome!

File: pkg/storage/chunk/cache/groupcache.go (hunk 3 above)
Comment (kavirajk) [reply pkg/storage/chunk/cache/groupcache.go]: Fixed :+1:

File: pkg/storage/chunk/cache/cache.go (hunk 2 above)
Comment (dannykopping) [reply pkg/storage/chunk/cache/cache.go:38]: Thanks for the explanation. I think we should resolve this somehow before merging

File: pkg/storage/chunk/cache/groupcache.go (hunk 6 above)
Comment (dannykopping) [reply pkg/storage/chunk/cache/groupcache.go:230]: Ah gotcha :+1: 
Thanks

File: pkg/storage/chunk/cache/cache.go (hunk 2 above)
Comment (kavirajk) [reply pkg/storage/chunk/cache/cache.go]: Fixed it in https://github.com/grafana/loki/pull/6793/commits/d7df061d90ae4ee5cb61963d3c781e152a1e7ebd

**Review by dannykopping — APPROVED** [review]
LGTM with a non-blocking nit

**Comment by kavirajk** [comment]
> <nit>Can we set config.CapacityMB * 1e6 in a single place?</nit>

I'm afraid we can't. Also both are in-fact different fields.  One is for global groupcache and other is for Local Group, both getting initialized at different places.

**Pull request merged**